components/eamxx/src/share/util/eamxx_time_interpolation.hpp

```diff
@@ -0,0 +1,86 @@
+#ifndef EAMXX_TIME_INTERPOLATION_HPP
+#define EAMXX_TIME_INTERPOLATION_HPP
+
+#include "share/grid/abstract_grid.hpp"
+
+#include "share/util/scream_time_stamp.hpp"
+
+#include "share/field/field.hpp"
+#include "share/field/field_manager.hpp"
+
+#include "share/io/scorpio_input.hpp"
+
+namespace scream{
+namespace util {
+
+class TimeInterpolation {
+public:
+   using grid_ptr_type = std::shared_ptr<const AbstractGrid>;
+   using vos_type = std::vector<std::string>;
+   using fm_type = std::shared_ptr<FieldManager>;
+
+  // Constructors
+  TimeInterpolation() = default;
+  TimeInterpolation(const grid_ptr_type& grid);
+  TimeInterpolation(const grid_ptr_type& grid, const vos_type& list_of_files);
+
+  // Running the interpolation
+  void initialize_timestamps(const TimeStamp& ts_in);
+  void initialize_data_from_field(const Field& field_in);
+  void initialize_data_from_files();
+  void update_data_from_field(const Field& field_in);
+  void update_timestamp(const TimeStamp& ts_in);
+  std::map<std::string,Field> perform_time_interpolation(const TimeStamp& time_in);
+
+  // Build interpolator
+  void add_field(const Field& field_in);
+
+  // Informational
+  void print();
+
+protected:
+
+  // Internal structure to store data source triplets (when using data from file)
+  // For each timesnap of data we have access to this triplet stores the
+  //  - filename
+  //  - timestamp
+  //  - time index in the file.
+  // Note, in many cases we will have files with multiple snaps of data and we
+  // need a good way to organize this information.
+  struct DataFromFileTriplet {
+  public:
+    std::string filename;
+    TimeStamp   timestamp;
+    int         time_idx;
+  };
+
+  // Helper functions to shift data
+  void shift_data();
+  void shift_data(const std::string& name);
+
+  // For the case where forcing data comes from files
+  void set_file_data_triplets(const vos_type& list_of_files);
+  void read_data();
+  void check_and_update_data(const TimeStamp& ts_in);
+
+  // Local field managers used to store two time snaps of data for interpolation
+  fm_type  m_fm_time0;
+  fm_type  m_fm_time1;
+  vos_type m_field_names;
+
+  // Store the timestamps associated with the two time snaps
+  TimeStamp m_time0;
+  TimeStamp m_time1;
+
+  // Variables related to the case where we use data from file
+  std::vector<DataFromFileTriplet>           m_file_data_triplets;
+  std::vector<DataFromFileTriplet>::iterator m_triplet_iterator;
```

Comment:
Doesn't look like you need to store this in the class

Reply:
This may be an incorrect usage of an iterator, if so I can change how it is used elsewhere, but the use here is to store the location in the `m_file_data_triplets` vector that is the most recent data snap we read.  So at any given time in the simulation `m_triplet_iterator` is pointing to a different element of the vector, and being updated.

Alternatively (and in a previous version) I had an int that stored the index of the vector and would update that.  And then grab the element of `m_file_data_triplets` using the index.